lib/ukcpio/cpio.c

```diff
@@ -292,6 +294,36 @@ read_section(struct cpio_header **header_ptr,
 			error = -UKCPIO_MKDIR_FAILED;
 			goto out;
 		}
+	} else if (IS_SYMLINK(header_mode)) {
+		uk_pr_info("Creating symlink %s\n", path_from_root);
+
+		data_location = (char *)ALIGN_4(
+				(char *)(header) + sizeof(struct cpio_header)
+				+ header_namesize);
+
+		if ((uintptr_t)data_location + header_filesize > last) {
+			uk_pr_err("%s: File exceeds archive bounds\n",
+				  path_from_root);
+			*header_ptr = NULL;
+			error = -UKCPIO_MALFORMED_INPUT;
+			goto out;
+		}
+
+		char target[header_filesize + 1];
+		memcpy(target, data_location, header_filesize);
+		target[header_filesize] = 0;
+
+		uk_pr_info("%s: Target is %s\n", path_from_root, target);
+		if (symlink(target, path_from_root)) {
```

Comment:
The rest of the file uses stdlib-like file access functions like [open](https://github.com/unikraft/unikraft/pull/992/files#diff-16c7dcd3c3387d5b8d30d83b32014a1bd9f4354b64f2deb66bfde8e9eec83cccR237), [write](https://github.com/unikraft/unikraft/pull/992/files#diff-16c7dcd3c3387d5b8d30d83b32014a1bd9f4354b64f2deb66bfde8e9eec83cccR263), [chmod](https://github.com/unikraft/unikraft/pull/992/files#diff-16c7dcd3c3387d5b8d30d83b32014a1bd9f4354b64f2deb66bfde8e9eec83cccR276), etc., and I assume that's a design decision to not depend on syscall-shim. I kept the same convention for the symlink code.

Reply:
The problem with libc-style syscalls is that you will exit the "kernel space" code and land into an external libc, that you can not control. There were some issues when this would lead to an infinite loop and other problems (see [this pr](https://github.com/unikraft/unikraft/pull/633)), so we decided to use raw syscalls from whithin the Unikraft core, so the execution will not leave the core space.

The `uk_syscall_*` symbols are defined even if the `syscall-shim` is not enabled, so most likely the rest of the file was written before we ran into issues with the libc-style syscalls.

Reply:
Right, I'm aware of the dangers of calling into libc from critical code but hadn't known about the raw syscall policy, nor that this file doesn't necessarily use libcalls on purpose; thanks for clarifying!

I'll use syscalls for symlink in this PR, and add fixing up the rest of the file (or module (or unikraft)) to my backlog to be handled in a future PR.